drivers/modem/modem_cellular.c

```diff
@@ -0,0 +1,1226 @@
+		return "script failed";
+	case MODEM_CELLULAR_EVENT_CMUX_CONNECTED:
+		return "cmux connected";
+	case MODEM_CELLULAR_EVENT_DLCI1_OPENED:
+		return "dlci1 opened";
+	case MODEM_CELLULAR_EVENT_DLCI2_OPENED:
+		return "dlci2 opened";
+	case MODEM_CELLULAR_EVENT_TIMEOUT:
+		return "timeout";
+	}
+
+	return "";
+}
+
+static bool modem_cellular_gpio_is_enabled(const struct gpio_dt_spec *gpio)
+{
+	return gpio->port != NULL;
+}
+
+static void modem_cellular_enter_state(struct modem_cellular_data *data,
+				       enum modem_cellular_state state);
+
+static void modem_cellular_delegate_event(struct modem_cellular_data *data,
+					  enum modem_cellular_event evt);
+
+static void modem_cellular_event_handler(struct modem_cellular_data *data,
+					 enum modem_cellular_event evt);
+
+static void modem_cellular_dlci1_pipe_handler(struct modem_pipe *pipe,
+					      enum modem_pipe_event event,
+					      void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	switch (event) {
+	case MODEM_PIPE_EVENT_OPENED:
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_DLCI1_OPENED);
+
+		break;
+
+	default:
+		break;
+	}
+}
+
+static void modem_cellular_dlci2_pipe_handler(struct modem_pipe *pipe,
+					      enum modem_pipe_event event,
+					      void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	switch (event) {
+	case MODEM_PIPE_EVENT_OPENED:
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_DLCI2_OPENED);
+
+		break;
+
+	default:
+		break;
+	}
+}
+
+static void modem_cellular_chat_callback_handler(struct modem_chat *chat,
+						 enum modem_chat_script_result result,
+						 void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (result == MODEM_CHAT_SCRIPT_RESULT_SUCCESS) {
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS);
+	} else {
+		modem_cellular_delegate_event(data, MODEM_CELLULAR_EVENT_SCRIPT_FAILED);
+	}
+}
+
+static void modem_cellular_chat_on_imei(struct modem_chat *chat, char **argv, uint16_t argc,
+					void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 2) {
+		return;
+	}
+
+	if (strlen(argv[1]) != 15) {
+		return;
+	}
+
+	for (uint8_t i = 0; i < 15; i++) {
+		data->imei[i] = argv[1][i] - '0';
+	}
+}
+
+static void modem_cellular_chat_on_cgmm(struct modem_chat *chat, char **argv, uint16_t argc,
+					void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 2) {
+		return;
+	}
+
+	strncpy(data->hwinfo, argv[1], sizeof(data->hwinfo) - 1);
+}
+
+static void modem_cellular_chat_on_creg(struct modem_chat *chat, char **argv, uint16_t argc,
+					void *user_data)
+{
+	struct modem_cellular_data *data = (struct modem_cellular_data *)user_data;
+
+	if (argc != 3) {
+		return;
+	}
+
+	data->access_tech = atoi(argv[1]);
```

Comment:
This is supposed to be the `AcT` value? This should be argv[5] then.

Reply:
I used this [resource](https://m2msupport.net/m2msupport/atcreg-network-registration/) as a reference, I can double check but I'm pretty sure this fits what the ublox and quectel modems respond with :)

Reply:
The key thing to notice is that the +CREG unsolicited event format is different than the queried response format on the HL7800. 
Based on your link, those modules work different and the command format is not standard which is very maddening.